File: packages/compiler/src/ir-to-dsl.ts
import { validateModelIR, type ModelIR, type SubjectRef, type ValidationError } from "@lazyfga/shared";
import { transformer } from "@openfga/syntax-transformer";
import { conditionToCel } from "./condition-to-cel";

/** OpenFGA AuthorizationModel JSON(공식 변환기 출력, id 제외). */
export type AuthModelJSON = ReturnType<typeof transformer.transformDSLToJSONObject>;

export class CompileError extends Error {
  constructor(
    public readonly reason: "IR_INVALID" | "JSON_TRANSFORM_FAILED",
    public readonly detail: unknown,
  ) {
    super(`compileIrToDsl failed: ${reason}`);
    this.name = "CompileError";
  }
}

/** SubjectRef → DSL 토큰. user | <group>#member (+ ` with <cond>` if conditioned). */
function serializeSubject(ref: SubjectRef): string {
  const base = ref.kind === "user" ? "user" : `${ref.group}#${ref.relation}`;
  return ref.condition !== undefined ? `${base} with ${ref.condition}` : base;
}

/** type restriction 직렬화: [a, b, ...] (IR 배열 순서 유지). */
function serializeSubjects(refs: SubjectRef[]): string {
  return `[${refs.map(serializeSubject).join(", ")}]`;
}

/**
 * ModelIR → OpenFGA DSL 문자열(결정적: 동일 IR → 바이트 단위 동일 DSL).
 * emit 순서: header → groups(입력순) → resources(입력순; relation은 parents→roles→permissions).
 */
function emitDsl(ir: ModelIR): string {
  const lines: string[] = ["model", "  schema 1.1", "type user"];

  for (const g of ir.groups) {
    lines.push(`type ${g.name}`, "  relations", `    define member: ${serializeSubjects(g.memberTypes)}`);
  }

  for (const r of ir.resources) {
    lines.push(`type ${r.name}`);
    const relLines: string[] = [];

    for (const p of r.parents) {
      relLines.push(`    define ${p.relationName}: [${p.parentTypes.join(", ")}]`);
    }
    for (const role of r.roles) {
      relLines.push(`    define ${role.name}: ${serializeSubjects(role.assignableBy)}`);
    }
    for (const perm of r.permissions) {
      const union = [
        ...perm.grantedByRoles,
        ...perm.inheritFromParents.map((rel) => `can_${perm.name} from ${rel}`),
      ].join(" or ");
      relLines.push(`    define can_${perm.name}: ${union}`);
    }

    if (relLines.length > 0) {
      lines.push("  relations", ...relLines);
    }
  }

  // 최상위 condition 블록(타입 뒤). 입력순 유지(결정적).
  for (const cond of ir.conditions ?? []) {
    const { decl, cel } = conditionToCel(cond);
    lines.push(`${decl} {`, `  ${cel}`, "}");
  }

  return lines.join("\n");
}

/**
 * ModelIR을 OpenFGA DSL 문자열과 AuthorizationModel JSON으로 컴파일한다.
 * 결정적: 동일 IR → 바이트 단위 동일 DSL. 호출 전 validateModelIR 통과 전제이며,
 * 방어적으로 재검증한다.
 * @throws CompileError IR 검증 위반 또는 JSON 변환 실패 시
 */
export function compileIrToDsl(ir: ModelIR): { dsl: string; model: AuthModelJSON } {
  const errors: ValidationError[] = validateModelIR(ir);
  if (errors.length > 0) {
    throw new CompileError("IR_INVALID", errors);
  }

  const dsl = emitDsl(ir);

  let model: AuthModelJSON;
  try {
    model = transformer.transformDSLToJSONObject(dsl);
  } catch (cause) {
    throw new CompileError("JSON_TRANSFORM_FAILED", cause);
  }

  return { dsl, model };
}

File: apps/api/internal/compiler/ir_to_dsl.go
package compiler

import (
	"strings"

	"github.com/antegral-constellation/lazyfga/api/internal/contract"
	"github.com/openfga/language/pkg/go/transformer"
)

// CompileError는 컴파일 실패 원인을 담는다. Reason: "IR_INVALID"(방어적 재검증 실패) |
// "JSON_TRANSFORM_FAILED"(공식 transformer가 emit된 DSL을 거부).
type CompileError struct {
	Reason string
	Detail any
}

func (e *CompileError) Error() string { return "compileIrToDsl failed: " + e.Reason }

// serializeSubject: SubjectRef → DSL 토큰. user | <group>#member (+ " with <cond>" if conditioned).
func serializeSubject(ref contract.SubjectRef) string {
	base := "user"
	if ref.Kind != "user" {
		base = ref.Group + "#" + ref.Relation
	}
	if ref.Condition != nil {
		return base + " with " + *ref.Condition
	}
	return base
}

// serializeSubjects: type restriction 직렬화 [a, b, ...](IR 배열 순서 유지).
func serializeSubjects(refs []contract.SubjectRef) string {
	parts := make([]string, len(refs))
	for i, r := range refs {
		parts[i] = serializeSubject(r)
	}
	return "[" + strings.Join(parts, ", ") + "]"
}

// emitDsl: ModelIR → OpenFGA DSL 문자열(결정적). emit 순서: header → groups(입력순) →
// resources(입력순; relation은 parents→roles→permissions) → 최상위 condition 블록(입력순).
func emitDsl(ir *contract.ModelIR) string {
	lines := []string{"model", "  schema 1.1", "type user"}

	for _, g := range ir.Groups {
		lines = append(lines,
			"type "+g.Name,
			"  relations",
			"    define member: "+serializeSubjects(g.MemberTypes),
		)
	}

	for _, r := range ir.Resources {
		lines = append(lines, "type "+r.Name)
		var relLines []string

		for _, p := range r.Parents {
			relLines = append(relLines, "    define "+p.RelationName+": ["+strings.Join(p.ParentTypes, ", ")+"]")
		}
		for _, role := range r.Roles {
			relLines = append(relLines, "    define "+role.Name+": "+serializeSubjects(role.AssignableBy))
		}
		for _, perm := range r.Permissions {
			union := make([]string, 0, len(perm.GrantedByRoles)+len(perm.InheritFromParents))
			union = append(union, perm.GrantedByRoles...)
			for _, rel := range perm.InheritFromParents {
				union = append(union, "can_"+perm.Name+" from "+rel)
			}
			relLines = append(relLines, "    define can_"+perm.Name+": "+strings.Join(union, " or "))
		}

		if len(relLines) > 0 {
			lines = append(lines, "  relations")
			lines = append(lines, relLines...)
		}
	}

	// 최상위 condition 블록(타입 뒤). 입력순 유지(결정적).
	for i := range ir.Conditions {
		decl, cel := ConditionToCel(&ir.Conditions[i])
		lines = append(lines, decl+" {", "  "+cel, "}")
	}

	return strings.Join(lines, "\n")
}

// CompileIRToDSL은 ModelIR을 결정적 .fga DSL 문자열과 AuthorizationModel JSON(공식
// openfga/language transformer 경유)으로 컴파일한다. 호출 전 ValidateModelIR 통과 전제이며
// 방어적으로 재검증한다. 실패 시 *CompileError를 반환한다.
func CompileIRToDSL(ir *contract.ModelIR) (dsl string, modelJSON []byte, err error) {
	if errs := contract.ValidateModelIR(ir); len(errs) > 0 {
		return "", nil, &CompileError{Reason: "IR_INVALID", Detail: errs}
	}

	dsl = emitDsl(ir)

	jsonStr, terr := transformer.TransformDSLToJSON(dsl)
	if terr != nil {
		return "", nil, &CompileError{Reason: "JSON_TRANSFORM_FAILED", Detail: terr}
	}

	return dsl, []byte(jsonStr), nil
}
